DirectProgramming/DPC++FPGA/ReferenceDesigns/mvdr_beamforming/src/mvdr_beamforming.cpp

```diff
@@ -0,0 +1,627 @@
+    data_out_consumer_event.wait();
+
+    one_passed = CheckOutputData(in_dir, DataOutConsumer::Data(), true);
+    if (one_passed) {
+      std::cout << "Output data check succeeded" << std::endl;
+    } else {
+      std::cout << "Output data check failed" << std::endl;
+    }
+    passed &= one_passed;
+
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Throughput test
+    ////////////////////////////////////////////////////////////////////////////
+    std::cout << std::endl
+              << "*** Launching throughput test of " << num_matrix_copies
+              << " matrices ***" << std::endl;
+  
+    std::cout << "Sensor inputs                 : " << kNumSensorInputs
+              << std::endl;
+    std::cout << "Training matrix rows          : " << kTrainingMatrixNumRows
+              << std::endl;
+    std::cout << "Data rows per training matrix : " << kNumInputVectors
+              << std::endl;
+    std::cout << "Steering vectors              : " << kNumSteer
+              << std::endl;
+
+    // start the consumer
+    data_out_consumer_event =
+        DataOutConsumer::Start(q, kDataOutSize * num_matrix_copies);
+
+    auto start_time = high_resolution_clock::now();
+
+    // start the producers
+    xrx_training_producer_event =
+        XrxTrainingProducer::Start(q, kTrainingDataSize * num_matrix_copies);
+    xrx_data_producer_event =
+        XrxDataProducer::Start(q, kInputDataSize * num_matrix_copies);
+
+    // wait for producer and consumer kernels to finish
+    xrx_training_producer_event.wait();
+    xrx_data_producer_event.wait();
+    data_out_consumer_event.wait();
+
+    auto end_time = high_resolution_clock::now();
+
+    // compute latency and throughput
+    duration<double, std::milli> process_time = end_time - start_time;
+    double latency_s = process_time.count() / 1000.0;
+    double throughput = (double)num_matrix_copies / latency_s;
+
+    std::cout << "Throughput: " << throughput << " matrices/second" << std::endl;
+
+    // check one instance of output data
+    one_passed = CheckOutputData(in_dir, DataOutConsumer::Data(), true);
+    if (one_passed) {
+      std::cout << "Output data check succeeded" << std::endl;
+    } else {
+      std::cout << "Output data check failed" << std::endl;
+    }
+    passed &= one_passed;
+
+    ////////////////////////////////////////////////////////////////////////////
+    // Post-test cleanup
+    ////////////////////////////////////////////////////////////////////////////
+
+    // free the host USM memory
+    XrxTrainingProducer::Destroy(q);
+    XrxDataProducer::Destroy(q);
+    SinThetaProducer::Destroy(q);
+    DataOutConsumer::Destroy(q);
+
+  } catch (exception const &e) {
+    // Catches exceptions in the host code
+    std::cerr << "Caught a SYCL host exception:\n" << e.what() << "\n";
+    // Most likely the runtime couldn't find FPGA hardware!
+    if (e.get_cl_code() == CL_DEVICE_NOT_FOUND) {
```

Comment:
Is there a DPC++ mechanism here so we don't have to drop to OpenCL?

Reply:
This is boilder plate code from all our reference designs.  I believe the only way to determine the device not found status (which is what users are most likely to run into here, so it warrants special handling) is by checking the OpenCL  code.